great_expectations/experimental/datasources/interfaces.py

```diff
@@ -57,6 +61,40 @@ class BatchRequestError(Exception):
     pass
 
 
+@pydantic_dc.dataclass(frozen=True)
+class BatchSorter:
+    key: str
+    reverse: bool = False
+
+
+BatchSortersDefinition: TypeAlias = Union[List[BatchSorter], List[str]]
+
+
+def _batch_sorter_from_list(sorters: BatchSortersDefinition) -> List[BatchSorter]:
+    if len(sorters) == 0 or isinstance(sorters[0], BatchSorter):
+        # mypy gets confused here. Since BatchSortersDefinition has all elements of the
+        # same type in the list so if the first on is BatchSorter so are the others.
+        return cast(List[BatchSorter], sorters)
+    # Likewise, sorters must be List[str] here.
+    return [_batch_sorter_from_str(sorter) for sorter in cast(List[str], sorters)]
+
+
+def _batch_sorter_from_str(sort_key: str) -> BatchSorter:
```

Comment:
```suggestion
def _batch_sorter_from_str(sort_key: Union[str, BatchSorter]) -> BatchSorter:
```
If you do this can you remove the `cast` on line `79`?

Reply:
If I do this, there are other mypy errors in `_batch_sorter_from_str` since it assumes the parameter is a `str`. I can get rid of casting by using `TypeGuard`s which I have done. It might be cleanest to use 1 `TypeGuard` and 1 `cast`. The `TypeGuard` can determine which one of the possible `Union` values it is but the `else` doesn't infer it must be the other `Union` type so I have to explicitly check for this.